R/route.R

```diff
@@ -99,28 +99,37 @@ route_dodgr <-
   }
 
   # Try to get route network if net not provided
-  pts <- rbind(from_coords, to_coords)
   if(is.null(net)) {
-    net <- dodgr::dodgr_streetnet(pts = pts, expand = 0.2)
+      pts <- rbind(from_coords, to_coords)
+      net <- dodgr::dodgr_streetnet(pts = pts, expand = 0.2)
   }
 
-  suppressWarnings(
-    suppressMessages(
-      ways_dg <- dodgr::weight_streetnet(net)
-      )
+  suppressMessages (
+    ways_dg <- dodgr::weight_streetnet(net)
   )
 
   verts <- dodgr::dodgr_vertices(ways_dg) # the vertices or points for routing
-  knf <- nabor::knn(cbind(verts$x, verts$y), matrix(from_coords, ncol = 2), k = 1)
-  knt <- nabor::knn(cbind(verts$x, verts$y), matrix(to_coords, ncol = 2), k = 1)
-  dp <- dodgr::dodgr_paths(ways_dg, from = verts$id[knf$nn.idx], to = verts$id[knt$nn.idx])
-  path <- verts[match(dp[[1]][[1]], verts$id), ]
-  path_linestring <- sf::st_linestring(cbind(path$x, path$y))
-  path_sf <- sf::st_sf(sf::st_sfc(path_linestring), crs = 4326)
-  # plot(net$geometry) # in there for the fun of it (to be removed)
-  # plot(path_linestring, col = "red", lwd = 5, add = TRUE)
-  path_sf
-
-  # clever routing code
-
+  #suppressMessages ({
+    from_id <- verts$id[dodgr::match_pts_to_graph(verts, from_coords)]
+    to_id <- verts$id[dodgr::match_pts_to_graph(verts, to_coords)]
+  #})
+  dp <- dodgr::dodgr_paths(ways_dg, from = from_id, to = to_id)
+  paths <- lapply(dp, function (i)
+                   lapply(i, function (j) {
+                             res <- verts[match (j, verts$id), c("x", "y")]
+                             sf::st_linestring(as.matrix(res))
+    }))
+  nms <- unlist(lapply(paths, function (i) names (i)))
+  from_to <- do.call(rbind, strsplit(nms, "-"))
```

Comment:
Very clever, still not 100% what's going on here, but seems to work!

Reply:
It's only intricate because `paths` is a list of lists, the second level of which has names formed from mashing together `"<from-id>-<to_id>"`. `nms` extracts that bit as one flat list; `strsplit` converts each list item into a 2-element vector, then `rbind` turns that into a 2-column matrix of `[from, to]`.